### PR TITLE
RasterDataset: fix support for flipped or upside down rasters

### LINKED ISSUE #1746: RasterDataset: no support for images with positive y-resolution
### Description

Certain geotifs have their axes flipped such as [this one](http://umbra-open-data-catalog.s3.amazonaws.com/sar-data/tasks/ad%20hoc/Sebarok%20Island%2C%20Singapore/6f073ed1-efdd-499f-bfdc-c102b7daf307/2023-06-23-03-23-26_UMBRA-06/2023-06-23-03-23-26_UMBRA-06_GEC.tif), which has the following transform flipping the direction of the x-axis:
`Affine(-0.346241067933152, 0.2915401535597796, 366317.96341952495, 0.2915401535597782, 0.3462410679331555, 130406.04105973356)`.

Depending on the transform, I believe this affects how random bounding boxes bounds should be ordered . For the example geotif above with x-axis flipped, [this line](https://github.com/microsoft/torchgeo/blob/4cc0cbb3276d4aae614a3c12381ad9b1e6a53292/torchgeo/datasets/geo.py#L536) maybe should be `bounds = (query.maxx, query.miny, query.minx, query.maxy)` instead of `bounds = (query.minx, query.miny, query.maxx, query.maxy)`. I have not tested it myself.
 
RandomGeoSampler and RasterDataset do not check for this, resulting in a `Bounds and transform are inconsistent` error

### Steps to reproduce

1. Download the sample data from [here](http://umbra-open-data-catalog.s3.amazonaws.com/sar-data/tasks/ad%20hoc/Sebarok%20Island%2C%20Singapore/6f073ed1-efdd-499f-bfdc-c102b7daf307/2023-06-23-03-23-26_UMBRA-06/2023-06-23-03-23-26_UMBRA-06_GEC.tif)
2. Run the following code to reproduce
 ```python
from torch.utils.data import DataLoader
from torchgeo.datasets import RasterDataset
from torchgeo.samplers import RandomGeoSampler
    
    
class UmbraDataset(RasterDataset):

    # umbra tifs are named like: 2023-08-02-17-08-40_UMBRA-05_GEC.tif
    filename_glob = '*.tif'
    filename_regex = r"^(?P<date>\d{4}-\d{2}-\d{2}-\d{2}-\d{2}-\d{2})"
    date_format = "%Y-%m-%d-%H-%M-%S"
    is_image = True
    separate_files = False 
    

dataset = UmbraDataset('/path/to/data/')

sampler = RandomGeoSampler(dataset, size=512, length=None)
dataloader = DataLoader(dataset, sampler=sampler, batch_size=2)

for i, batch in enumerate(dataloader):
    sample = batch
```
### Version

0.5.1

### PR DESCRIPTION
**Pull request by adamjstewart**
Context: https://github.com/rasterio/rasterio/issues/3094
Fixes #1746 
Required by #3244

### CODE REVIEW

**Comment by calebrob6** [comment]
Can you add tests for this? Lack of coverage for the negative x and y dimension cases are hid by the fact that its all on one line

**Comment by adamjstewart** [comment]
I tried updating `tests/data/raster/data.py` to produce flipped and upside down images but couldn't get it working. #3244 does test upside down images, but I would love to test all possible combinations. Need to try again.

**Comment by adamjstewart** [comment]
Tests revealed that my fix works for upside down rasters (the ones we care most about) but not for flipped rasters. Will play around a bit more, but may give up and focus on upside down rasters.
